wire-schema/src/main/java/com/squareup/wire/schema/EnumConstant.java

```diff
@@ -13,78 +13,35 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.squareup.wire.schema;
-
-import com.google.common.collect.ImmutableList;
-import com.squareup.wire.schema.internal.parser.EnumConstantElement;
-import java.util.List;
-
-public final class EnumConstant {
-  private final Location location;
-  private final String name;
-  private final int tag;
-  private final String documentation;
-  private final Options options;
-
-  private EnumConstant(Location location, String name, int tag, String documentation,
-      Options options) {
-    this.location = location;
-    this.name = name;
-    this.tag = tag;
-    this.documentation = documentation;
-    this.options = options;
-  }
-
-  public Location location() {
-    return location;
-  }
-
-  public String name() {
-    return name;
-  }
-
-  public int tag() {
-    return tag;
-  }
-
-  public String documentation() {
-    return documentation;
-  }
-
-  public Options options() {
-    return options;
-  }
-
-  static EnumConstant fromElement(EnumConstantElement element) {
-    return new EnumConstant(element.getLocation(), element.getName(), element.getTag(),
-        element.getDocumentation(), new Options(Options.ENUM_VALUE_OPTIONS, element.getOptions()));
-  }
-
-  EnumConstantElement toElement() {
-    return new EnumConstantElement(location, name, tag, documentation, options.toElements());
-  }
-
-  void linkOptions(Linker linker) {
-    options.link(linker);
-  }
-
-  EnumConstant retainAll(Schema schema, MarkSet markSet) {
-    return new EnumConstant(location, name, tag, documentation, options.retainAll(schema, markSet));
-  }
-
-  static ImmutableList<EnumConstant> fromElements(List<EnumConstantElement> elements) {
-    ImmutableList.Builder<EnumConstant> constants = ImmutableList.builder();
-    for (EnumConstantElement element : elements) {
-      constants.add(fromElement(element));
-    }
-    return constants.build();
-  }
-
-  static ImmutableList<EnumConstantElement> toElements(List<EnumConstant> constants) {
-    ImmutableList.Builder<EnumConstantElement> elements = new ImmutableList.Builder<>();
-    for (EnumConstant constant : constants) {
-      elements.add(constant.toElement());
-    }
-    return elements.build();
+package com.squareup.wire.schema
+
+import com.squareup.wire.schema.internal.parser.EnumConstantElement
+
+class EnumConstant private constructor(
+  val location: Location,
+  val name: String,
+  val tag: Int,
+  val documentation: String,
+  val options: Options
+) {
+  internal fun toElement() =
+    EnumConstantElement(location, name, tag, documentation, options.toElements())
+
+  internal fun linkOptions(linker: Linker) = options.link(linker)
+  internal fun retainAll(
```

Comment:
We got rid of this kind of methods in KotlinPoet in favor of data classes and `copy()`, that's a more idiomatic approach to API. We should into doing the same everywhere in Wire.

Reply:
Good follow-up